static/src/javascripts/projects/common/modules/commercial/track-ad.js

```diff
@@ -0,0 +1,41 @@
+define([
+    'Promise',
+    'common/utils/mediator'
+], function (
+    Promise,
+    mediator
+) {
+    var promises;
+
+    function TrackAd() {
+        promises = {};
+    }
+
+    TrackAd.prototype.track = function (id) {
+        promises[id] = promises[id] || new Promise(function (resolve, reject) {
+            var onAdLoaded = function (event) {
+                if (event.slot.getSlotElementId() === id) {
+                    unlisten();
+                    resolve(!event.isEmpty);
+                }
+            };
+
+            var onAllAdsLoaded = function () {
+                unlisten();
+                reject(new Error('Slot ' + id + ' was never loaded'));
+            };
+
+            function unlisten() {
+                mediator.off('modules:commercial:dfp:rendered', onAdLoaded);
+                mediator.off('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+            }
+
+            mediator.on('modules:commercial:dfp:rendered', onAdLoaded);
+            mediator.on('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+        });
+
+        return promises[id];
+    };
+
+    return new TrackAd();
```

Comment:
The initial version was a function. I'm using the constructor form so that the cache can be reset during testing. This is not great since I am against any use of classes in JS, but it's that or adding a reset function. Considering a second parameter `force` that would be false by default though.

Thoughts?